File: app/src/main/java/com/tablet/hid/util/GamepadConfigSerializer.kt
package com.tablet.hid.util

import com.tablet.hid.model.ButtonConfig
import com.tablet.hid.model.ClickBehavior
import com.tablet.hid.model.GamepadConfig
import com.tablet.hid.model.JoystickConfig
import com.tablet.hid.model.JoystickSide
import com.tablet.hid.model.KeyboardMacroButtonConfig
import com.tablet.hid.model.MacroHostDefaults
import com.tablet.hid.model.TriggerDragAxis
import com.tablet.hid.model.VibrationIntensity
import org.json.JSONArray
import org.json.JSONObject

object GamepadConfigSerializer {

    private val BUTTON_KEYS = listOf(
        "a", "b", "x", "y", "lb", "rb", "lt", "rt",
        "back", "start", "dpadUp", "dpadDown", "dpadLeft", "dpadRight",
    )

    private val TRIGGER_KEYS = setOf("lt", "rt")

    fun toCanonicalJson(config: GamepadConfig): JSONObject {
        val root = JSONObject()
        val buttons = JSONObject()
        buttons.put("a",         buttonToJson(config.btnA))
        buttons.put("b",         buttonToJson(config.btnB))
        buttons.put("x",         buttonToJson(config.btnX))
        buttons.put("y",         buttonToJson(config.btnY))
        buttons.put("lb",        buttonToJson(config.btnLb))
        buttons.put("rb",        buttonToJson(config.btnRb))
        buttons.put("lt",        buttonToJson(config.btnLt))
        buttons.put("rt",        buttonToJson(config.btnRt))
        buttons.put("back",      buttonToJson(config.btnBack))
        buttons.put("start",     buttonToJson(config.btnStart))
        buttons.put("dpadUp",    buttonToJson(config.dpadUp))
        buttons.put("dpadDown",  buttonToJson(config.dpadDown))
        buttons.put("dpadLeft",  buttonToJson(config.dpadLeft))
        buttons.put("dpadRight", buttonToJson(config.dpadRight))
        root.put("buttons", buttons)
        root.put("leftJoystick",  joystickToJson(config.leftJoystick))
        root.put("rightJoystick", joystickToJson(config.rightJoystick))
        root.put("singleJoystickMode", config.singleJoystickMode)
        root.put("singleJoystickSideToggleEnabled", config.singleJoystickSideToggleEnabled)
        root.put("singleJoystickOutputSide", config.singleJoystickOutputSide.name)
        root.put("macroHostDefaults", config.macroHostDefaults.name)
        root.put("macroButtons", macrosToJson(config.macroButtons))
        root.put("vibrationIntensity", config.vibrationIntensity.name)
        val labels = JSONObject()
        config.customButtonLabels.forEach { (k, v) -> labels.put(k, v) }
        root.put("customButtonLabels", labels)
        return root
    }

    fun fromCanonicalJson(json: JSONObject): GamepadConfig {
        val buttons = json.optJSONObject("buttons") ?: JSONObject()
        val leftJs  = json.optJSONObject("leftJoystick")
        val rightJs = json.optJSONObject("rightJoystick")
        return GamepadConfig(
            btnA     = buttonFromJson(buttons.optJSONObject("a"),         isTrigger = false),
            btnB     = buttonFromJson(buttons.optJSONObject("b"),         isTrigger = false),
            btnX     = buttonFromJson(buttons.optJSONObject("x"),         isTrigger = false),
            btnY     = buttonFromJson(buttons.optJSONObject("y"),         isTrigger = false),
            btnLb    = buttonFromJson(buttons.optJSONObject("lb"),        isTrigger = false),
            btnRb    = buttonFromJson(buttons.optJSONObject("rb"),        isTrigger = false),
            btnLt    = buttonFromJson(buttons.optJSONObject("lt"),        isTrigger = true),
            btnRt    = buttonFromJson(buttons.optJSONObject("rt"),        isTrigger = true),
            btnBack  = buttonFromJson(buttons.optJSONObject("back"),      isTrigger = false),
            btnStart = buttonFromJson(buttons.optJSONObject("start"),     isTrigger = false),
            dpadUp    = buttonFromJson(buttons.optJSONObject("dpadUp"),    isTrigger = false),
            dpadDown  = buttonFromJson(buttons.optJSONObject("dpadDown"),  isTrigger = false),
            dpadLeft  = buttonFromJson(buttons.optJSONObject("dpadLeft"),  isTrigger = false),
            dpadRight = buttonFromJson(buttons.optJSONObject("dpadRight"), isTrigger = false),
            leftJoystick  = joystickFromJson(leftJs),
            rightJoystick = joystickFromJson(rightJs),
            singleJoystickMode = json.optBoolean("singleJoystickMode", false),
            singleJoystickSideToggleEnabled = json.optBoolean("singleJoystickSideToggleEnabled", false),
            singleJoystickOutputSide = enumValueOrDefault(
                json.optString("singleJoystickOutputSide"),
                JoystickSide.LEFT,
            ),
            macroHostDefaults = enumValueOrDefault(
                json.optString("macroHostDefaults"),
                MacroHostDefaults.WINDOWS,
            ),
            macroButtons = macrosFromJson(json.optJSONArray("macroButtons")),
            vibrationIntensity = enumValueOrDefault(
                json.optString("vibrationIntensity"),
                VibrationIntensity.OFF,
            ),
            customButtonLabels = labelsFromJson(json.optJSONObject("customButtonLabels")),
        )
    }

    private fun buttonToJson(btn: ButtonConfig): JSONObject {
        val j = JSONObject()
        j.put("enabled", btn.enabled)
        j.put("behavior", btn.behavior.name)
        j.put("turbo", btn.turbo)
        j.put("turboDurationMs", btn.turboDurationMs)
        j.put("turboIntervalMs", btn.turboIntervalMs)
        j.put("offsetX", btn.offsetX.toDouble())
        j.put("offsetY", btn.offsetY.toDouble())
        j.put("scaleX", btn.scaleX.toDouble())
        j.put("scaleY", btn.scaleY.toDouble())
        j.put("triggerTravelDp", btn.triggerTravelDp.toDouble())
        j.put("triggerAxis", btn.triggerAxis.name)
        return j
    }

    private fun buttonFromJson(j: JSONObject?, isTrigger: Boolean): ButtonConfig {
        val defaults = ButtonConfig()
        if (j == null) return defaults
        return ButtonConfig(
            enabled         = j.optBoolean("enabled", defaults.enabled),
            behavior        = enumValueOrDefault(j.optString("behavior"), defaults.behavior),
            turbo           = j.optBoolean("turbo", defaults.turbo),
            turboDurationMs = j.optInt("turboDurationMs", defaults.turboDurationMs),
            turboIntervalMs = j.optInt("turboIntervalMs", defaults.turboIntervalMs),
            offsetX         = j.optDouble("offsetX", defaults.offsetX.toDouble()).toFloat(),
            offsetY         = j.optDouble("offsetY", defaults.offsetY.toDouble()).toFloat(),
            scaleX          = j.optDouble("scaleX", defaults.scaleX.toDouble()).toFloat(),
            scaleY          = j.optDouble("scaleY", defaults.scaleY.toDouble()).toFloat(),
            triggerTravelDp = if (isTrigger) j.optDouble("triggerTravelDp", defaults.triggerTravelDp.toDouble()).toFloat()
                              else defaults.triggerTravelDp,
            triggerAxis     = if (isTrigger) enumValueOrDefault(j.optString("triggerAxis"), defaults.triggerAxis)
                              else defaults.triggerAxis,
        )
    }

    private fun joystickToJson(js: JoystickConfig): JSONObject {
        val j = JSONObject()
        j.put("enabled", js.enabled)
        j.put("deadzone", js.deadzone.toDouble())
        j.put("gain", js.gain.toDouble())
        j.put("offsetX", js.offsetX.toDouble())
        j.put("offsetY", js.offsetY.toDouble())
        j.put("scaleX", js.scaleX.toDouble())
        j.put("scaleY", js.scaleY.toDouble())
        return j
    }

    private fun joystickFromJson(j: JSONObject?): JoystickConfig {
        val defaults = JoystickConfig()
        if (j == null) return defaults
        return JoystickConfig(
            enabled  = j.optBoolean("enabled", defaults.enabled),
            deadzone = j.optDouble("deadzone", defaults.deadzone.toDouble()).toFloat(),
            gain     = j.optDouble("gain", defaults.gain.toDouble()).toFloat(),
            offsetX  = j.optDouble("offsetX", defaults.offsetX.toDouble()).toFloat(),
            offsetY  = j.optDouble("offsetY", defaults.offsetY.toDouble()).toFloat(),
            scaleX   = j.optDouble("scaleX", defaults.scaleX.toDouble()).toFloat(),
            scaleY   = j.optDouble("scaleY", defaults.scaleY.toDouble()).toFloat(),
        )
    }

    private fun macrosToJson(macros: List<KeyboardMacroButtonConfig>): JSONArray {
        val arr = JSONArray()
        macros.forEach { macro ->
            val j = JSONObject()
            j.put("label", macro.label)
            j.put("modifiers", macro.modifiers)
            val keys = JSONArray()
            macro.keyUsages.forEach { keys.put(it) }
            j.put("keyUsages", keys)
            j.put("layoutOffsetX", macro.layoutOffsetX.toDouble())
            j.put("layoutOffsetY", macro.layoutOffsetY.toDouble())
            j.put("layoutScaleX", macro.layoutScaleX.toDouble())
            j.put("layoutScaleY", macro.layoutScaleY.toDouble())
            arr.put(j)
        }
        return arr
    }

    private fun macrosFromJson(arr: JSONArray?): List<KeyboardMacroButtonConfig> {
        if (arr == null) return emptyList()
        return List(arr.length()) { i ->
            val j = arr.getJSONObject(i)
            val keysArr = j.optJSONArray("keyUsages")
            val keys = if (keysArr != null) List(keysArr.length()) { k -> keysArr.getInt(k) } else emptyList()
            KeyboardMacroButtonConfig(
                label         = j.optString("label", ""),
                modifiers     = j.optInt("modifiers", 0),
                keyUsages     = keys,
                layoutOffsetX = j.optDouble("layoutOffsetX", 0.0).toFloat(),
                layoutOffsetY = j.optDouble("layoutOffsetY", 0.0).toFloat(),
                layoutScaleX  = j.optDouble("layoutScaleX", 1.0).toFloat(),
                layoutScaleY  = j.optDouble("layoutScaleY", 1.0).toFloat(),
            )
        }.filter { it.label.isNotBlank() && it.keyUsages.isNotEmpty() }
    }

    private fun labelsFromJson(j: JSONObject?): Map<String, String> {
        if (j == null) return emptyMap()
        val map = mutableMapOf<String, String>()
        j.keys().forEach { key -> map[key] = j.getString(key) }
        return map
    }

    private inline fun <reified T : Enum<T>> enumValueOrDefault(name: String?, default: T): T =
        if (!name.isNullOrEmpty()) runCatching { enumValueOf<T>(name) }.getOrDefault(default) else default
}

File: ios/TabletHID/Services/GamepadConfigSerializer.swift
import Foundation

/// Converts GamepadConfig to and from the canonical JSON dictionary format defined in
/// spec/server-schema.md. Field names and nesting structure are identical to
/// GamepadConfigSerializer.kt on Android.
enum GamepadConfigSerializer {

    // MARK: - To canonical JSON

    static func toCanonicalJson(_ config: GamepadConfig) -> [String: Any] {
        var root: [String: Any] = [:]
        var buttons: [String: Any] = [:]
        buttons["a"]         = buttonToJson(config.btnA,     isTrigger: false)
        buttons["b"]         = buttonToJson(config.btnB,     isTrigger: false)
        buttons["x"]         = buttonToJson(config.btnX,     isTrigger: false)
        buttons["y"]         = buttonToJson(config.btnY,     isTrigger: false)
        buttons["lb"]        = buttonToJson(config.btnLB,    isTrigger: false)
        buttons["rb"]        = buttonToJson(config.btnRB,    isTrigger: false)
        buttons["lt"]        = buttonToJson(config.btnLT,    isTrigger: true)
        buttons["rt"]        = buttonToJson(config.btnRT,    isTrigger: true)
        buttons["back"]      = buttonToJson(config.btnBack,  isTrigger: false)
        buttons["start"]     = buttonToJson(config.btnStart, isTrigger: false)
        buttons["dpadUp"]    = buttonToJson(config.dpadUp,    isTrigger: false)
        buttons["dpadDown"]  = buttonToJson(config.dpadDown,  isTrigger: false)
        buttons["dpadLeft"]  = buttonToJson(config.dpadLeft,  isTrigger: false)
        buttons["dpadRight"] = buttonToJson(config.dpadRight, isTrigger: false)
        root["buttons"] = buttons
        root["leftJoystick"]  = joystickToJson(config.leftJoystick)
        root["rightJoystick"] = joystickToJson(config.rightJoystick)
        root["singleJoystickMode"]             = config.singleJoystickMode
        root["singleJoystickSideToggleEnabled"] = config.singleJoystickSideToggleEnabled
        root["singleJoystickOutputSide"] = config.singleJoystickOutputSide.rawValue.uppercased()
        root["macroHostDefaults"] = config.macroHostDefaults.rawValue.uppercased()
        root["macroButtons"] = config.macroButtons.map { macroToJson($0) }
        root["vibrationIntensity"] = config.vibrationIntensity.rawValue.uppercased()
        var labels: [String: String] = [:]
        config.customButtonLabels.forEach { labels[$0.key] = $0.value }
        root["customButtonLabels"] = labels
        return root
    }

    // MARK: - From canonical JSON

    static func fromCanonicalJson(_ json: [String: Any]) -> GamepadConfig {
        let buttons = json["buttons"] as? [String: Any] ?? [:]
        let leftJs  = json["leftJoystick"]  as? [String: Any]
        let rightJs = json["rightJoystick"] as? [String: Any]
        var config = GamepadConfig()
        config.btnA     = buttonFromJson(buttons["a"]         as? [String: Any], isTrigger: false)
        config.btnB     = buttonFromJson(buttons["b"]         as? [String: Any], isTrigger: false)
        config.btnX     = buttonFromJson(buttons["x"]         as? [String: Any], isTrigger: false)
        config.btnY     = buttonFromJson(buttons["y"]         as? [String: Any], isTrigger: false)
        config.btnLB    = buttonFromJson(buttons["lb"]        as? [String: Any], isTrigger: false)
        config.btnRB    = buttonFromJson(buttons["rb"]        as? [String: Any], isTrigger: false)
        config.btnLT    = buttonFromJson(buttons["lt"]        as? [String: Any], isTrigger: true)
        config.btnRT    = buttonFromJson(buttons["rt"]        as? [String: Any], isTrigger: true)
        config.btnBack  = buttonFromJson(buttons["back"]      as? [String: Any], isTrigger: false)
        config.btnStart = buttonFromJson(buttons["start"]     as? [String: Any], isTrigger: false)
        config.dpadUp    = buttonFromJson(buttons["dpadUp"]    as? [String: Any], isTrigger: false)
        config.dpadDown  = buttonFromJson(buttons["dpadDown"]  as? [String: Any], isTrigger: false)
        config.dpadLeft  = buttonFromJson(buttons["dpadLeft"]  as? [String: Any], isTrigger: false)
        config.dpadRight = buttonFromJson(buttons["dpadRight"] as? [String: Any], isTrigger: false)
        config.leftJoystick  = joystickFromJson(leftJs)
        config.rightJoystick = joystickFromJson(rightJs)
        config.singleJoystickMode             = json["singleJoystickMode"] as? Bool ?? false
        config.singleJoystickSideToggleEnabled = json["singleJoystickSideToggleEnabled"] as? Bool ?? false
        config.singleJoystickOutputSide = joystickSide(from: json["singleJoystickOutputSide"] as? String) ?? .left
        config.macroHostDefaults = macroHost(from: json["macroHostDefaults"] as? String) ?? .windows
        config.macroButtons = (json["macroButtons"] as? [[String: Any]] ?? []).compactMap { macroFromJson($0) }
        config.vibrationIntensity = vibrationIntensity(from: json["vibrationIntensity"] as? String) ?? .off
        config.customButtonLabels = json["customButtonLabels"] as? [String: String] ?? [:]
        return config
    }

    // MARK: - ButtonConfig helpers

    private static func buttonToJson(_ btn: ButtonConfig, isTrigger: Bool) -> [String: Any] {
        let j: [String: Any] = [
            "enabled":         btn.enabled,
            "behavior":        btn.behavior.rawValue.uppercased(),
            "turbo":           btn.turbo,
            "turboDurationMs": btn.turboDurationMs,
            "turboIntervalMs": btn.turboIntervalMs,
            "offsetX":         btn.offsetX,
            "offsetY":         btn.offsetY,
            "scaleX":          btn.scaleX,
            "scaleY":          btn.scaleY,
            "triggerTravelDp": btn.triggerTravel,
            "triggerAxis":     btn.triggerAxis.rawValue.uppercased(),
        ]
        return j
    }

    private static func buttonFromJson(_ j: [String: Any]?, isTrigger: Bool) -> ButtonConfig {
        var btn = ButtonConfig()
        guard let j = j else { return btn }
        btn.enabled         = j["enabled"]         as? Bool   ?? btn.enabled
        btn.behavior        = clickBehavior(from: j["behavior"] as? String) ?? btn.behavior
        btn.turbo           = j["turbo"]           as? Bool   ?? btn.turbo
        btn.turboDurationMs = j["turboDurationMs"] as? Int    ?? btn.turboDurationMs
        btn.turboIntervalMs = j["turboIntervalMs"] as? Int    ?? btn.turboIntervalMs
        btn.offsetX         = (j["offsetX"] as? Double) ?? btn.offsetX
        btn.offsetY         = (j["offsetY"] as? Double) ?? btn.offsetY
        btn.scaleX          = (j["scaleX"]  as? Double) ?? btn.scaleX
        btn.scaleY          = (j["scaleY"]  as? Double) ?? btn.scaleY
        if isTrigger {
            btn.triggerTravel = (j["triggerTravelDp"] as? Double) ?? btn.triggerTravel
            btn.triggerAxis   = triggerAxis(from: j["triggerAxis"] as? String) ?? btn.triggerAxis
        }
        return btn
    }

    // MARK: - JoystickConfig helpers

    private static func joystickToJson(_ js: JoystickConfig) -> [String: Any] {
        return [
            "enabled":  js.enabled,
            "deadzone": js.deadzone,
            "gain":     js.gain,
            "offsetX":  js.offsetX,
            "offsetY":  js.offsetY,
            "scaleX":   js.scaleX,
            "scaleY":   js.scaleY,
        ]
    }

    private static func joystickFromJson(_ j: [String: Any]?) -> JoystickConfig {
        var js = JoystickConfig()
        guard let j = j else { return js }
        js.enabled  = j["enabled"]  as? Bool   ?? js.enabled
        js.deadzone = (j["deadzone"] as? Double) ?? js.deadzone
        js.gain     = (j["gain"]     as? Double) ?? js.gain
        js.offsetX  = (j["offsetX"]  as? Double) ?? js.offsetX
        js.offsetY  = (j["offsetY"]  as? Double) ?? js.offsetY
        js.scaleX   = (j["scaleX"]   as? Double) ?? js.scaleX
        js.scaleY   = (j["scaleY"]   as? Double) ?? js.scaleY
        return js
    }

    // MARK: - Macro helpers

    private static func macroToJson(_ macro: KeyboardMacroButtonConfig) -> [String: Any] {
        return [
            "label":         macro.label,
            "modifiers":     macro.modifiers,
            "keyUsages":     macro.keyUsages,
            "layoutOffsetX": macro.layoutOffsetX,
            "layoutOffsetY": macro.layoutOffsetY,
            "layoutScaleX":  macro.layoutScaleX,
            "layoutScaleY":  macro.layoutScaleY,
        ]
    }

    private static func macroFromJson(_ j: [String: Any]) -> KeyboardMacroButtonConfig? {
        let label    = j["label"] as? String ?? ""
        let keyUsages = j["keyUsages"] as? [Int] ?? []
        guard !label.isEmpty, !keyUsages.isEmpty else { return nil }
        return KeyboardMacroButtonConfig(
            label:         label,
            modifiers:     j["modifiers"]     as? Int    ?? 0,
            keyUsages:     keyUsages,
            layoutOffsetX: j["layoutOffsetX"] as? Double ?? 0,
            layoutOffsetY: j["layoutOffsetY"] as? Double ?? 0,
            layoutScaleX:  j["layoutScaleX"]  as? Double ?? 1,
            layoutScaleY:  j["layoutScaleY"]  as? Double ?? 1
        )
    }

    // MARK: - Enum parsing (server uses UPPER_CASE, iOS model uses lower camelCase)

    private static func clickBehavior(from raw: String?) -> ClickBehavior? {
        switch raw?.uppercased() {
        case "MOMENTARY": return .momentary
        case "LATCHING":  return .latching
        default:          return nil
        }
    }

    private static func joystickSide(from raw: String?) -> JoystickSide? {
        switch raw?.uppercased() {
        case "LEFT":  return .left
        case "RIGHT": return .right
        default:      return nil
        }
    }

    private static func triggerAxis(from raw: String?) -> TriggerDragAxis? {
        switch raw?.uppercased() {
        case "UP":    return .up
        case "DOWN":  return .down
        case "LEFT":  return .left
        case "RIGHT": return .right
        default:      return nil
        }
    }

    private static func macroHost(from raw: String?) -> MacroHostDefaults? {
        switch raw?.uppercased() {
        case "WINDOWS": return .windows
        case "MAC":     return .mac
        default:        return nil
        }
    }

    private static func vibrationIntensity(from raw: String?) -> VibrationIntensity? {
        switch raw?.uppercased() {
        case "OFF":    return .off
        case "LIGHT":  return .light
        case "MEDIUM": return .medium
        case "STRONG": return .strong
        default:       return nil
        }
    }
}
